src/AI/JakeMan.java

```diff
@@ -0,0 +1,770 @@
+    path.snip(unit.getMovePower(gameMap) + 1); // Trim the path approximately down to size.
+    XYCoord pathPoint = path.getEndCoord(); // Set the last location as our goal.
+
+    // Sort my currently-reachable move locations by distance from the goal,
+    // and build a GameAction to move to the closest one.
+    Utils.sortLocationsByDistance(pathPoint, destinations);
+    log(String.format("  %s is traveling toward %s at %s via %s  mustMove?: %s",
+                          unit.toStringWithLocation(),
+                          gameMap.getLocation(goal).getEnvironment().terrainType, goal,
+                          pathPoint, mustMove));
+    for( XYCoord xyc : destinations )
+    {
+      log(String.format("    is it safe to go to %s?", xyc));
+      if( !isDudeFree(gameMap, unit, xyc, false) )
+        continue;
+
+      GameAction action = null;
+      Unit resident = gameMap.getLocation(xyc).getResident();
+      if( null != resident && unit != resident )
+      {
+        if( unit.CO == resident.CO && !resident.isTurnOver )
+          action = evictUnit(gameMap, unit, resident, avoidProduction);
+        if( null != action ) return action;
+        continue;
+      }
+      log(String.format("    Yes"));
+
+      GamePath movePath = Utils.findShortestPath(unit, xyc, gameMap);
+      ArrayList<GameActionSet> actionSets = unit.getPossibleActions(gameMap, movePath, ignoreResident);
+      if( actionSets.size() > 0 )
+      {
+        // Since we're moving anyway, might as well try shooting the scenery
+        for( GameActionSet actionSet : actionSets )
+        {
+          if( actionSet.getSelected().getType() == UnitActionFactory.ATTACK )
+          {
+            action = findBestAttack(gameMap, unit, actionSet.getGameActions());
+          }
+        }
+
+        if( null == action && movePath.getPathLength() > 1) // Just wait if we can't do anything cool
+          action = new WaitLifecycle.WaitAction(unit, movePath);
+        return action;
+      }
+    }
+    return null;
+  }
+
+  public static GameAction findBestAttack(GameMap gameMap, Unit unit, ArrayList<GameAction> actionSet)
+  {
+    double bestDamage = 0;
+    GameAction bestAttack = null;
+    for( GameAction attack : actionSet )
+    {
+      double damageValue = AICombatUtils.scoreAttackAction(unit, attack, gameMap,
+          (results) -> {
+            final Unit defender = results.defender.unit;
+            int loss   = Math.min(unit    .getHP(), (int)results.attacker.getPreciseHPDamage());
+            int damage = Math.min(defender.getHP(), (int)results.defender.getPreciseHPDamage());
+
+            // Convert to abstract value
+            int extraLoss = 0;
+            if( loss >= 1 && unit.getHP() == UnitModel.MAXIMUM_HP )
+              extraLoss += STAY_UNHURT_BIAS;
+            if( loss >= unit.getHP() )
+              extraLoss += STAY_ALIVE_BIAS;
+            loss *= unit.getCost();
+            loss += extraLoss;
```

Comment:
These loss bias factors will have a much greater influence over cheap unit interactions than expensive ones.
The bias for staying unhurt is as much as the cost of an infantry, but piddling in the context of a Md Tank. Should this factor be scaling wrt unit cost?

Reply:
That was the intent, actually.
An unhurt inf is *massively* more useful than a chipped one, because unhurt infs are very difficult to oneshot.
Tech units can generally still contribute well when hurt, though.